foundations/javascript_basics/understanding_errors.md

```diff
@@ -0,0 +1,132 @@
+
+# Common Types of Errors
+
+These are some of the most common errors you will encounter, so it’s important to understand them.
+
+## Syntax Error
+
+A syntax error occurs when the code you are trying to run is syntactically incorrect. For example this:
+
+```js
+function helloWorld() {
+console.log “Hello World!”
+}
+```
+
+will throw the following error, because we forgot the parenthesis for `console.log()`!
+<img width="602" alt="Screen Shot 2021-05-17 at 12 15 11 PM" src="https://user-images.githubusercontent.com/22967723/118543788-9823be00-b709-11eb-8f2a-3e4f64cee880.png">
+[MDN | SyntaxError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SyntaxError)
+
+## Reference Error
+
+We covered reference errors in the first example in this lesson, but it’s important to remember that these arise because whatever variable you are trying to reference does not exist.
+
+[MDN | ReferenceError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ReferenceError)
+
+## Type Error
+
+These errors are thrown for a few different reasons:
+
+Per MDN, a `TypeError` may be thrown when:
+
+> - an operand or argument passed to a function is incompatible with the type expected by that operator or function;
+> - or when attempting to modify a value that cannot be changed;
+> - or when attempting to use a value in an inappropriate way.
+
+Say we have two strings that you would like to combine to create one message such as below:
+
+```js
+const str1 = "Hello";
+const str2 = "World!";
+const message = str1.push(str2);
+```
+<img width="621" alt="Screen Shot 2021-05-17 at 12 14 12 PM" src="https://user-images.githubusercontent.com/22967723/118543650-6579c580-b709-11eb-981d-4bdb74966ba4.png">
+
+Here, we get a `TypeError` with a message stating that `str1.push is not a function`. This is a common error message that confuses learners, because we know that `.push()` is certainly a function! You’ve probably used it to add items to _arrays_ before, but that’s the key- `.push()` is not a String method, it’s an Array method. Hence, it is “not a function” that you can find as a String method. If we change `.push()` to `.concat()`, a proper String method, our code runs as intended!
+
+[MDN | TypeError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError)
+
+# Tips for Resolving Errors
+
+At this point, you might be wondering how we can resolve these errors.
+
+1. The first and most important step is to Google the error! Chances are, you can find a fix on StackOverflow or in the documentation. If anything, you will receive more clarity as to why you are receiving this error.
+2. Make use of the console! `console.log()` is a popular choice for quick debugging. For more involved troubleshooting, using the debugger might be more appropriate, but using `console.log()` is great for getting immediate feedback without needing to step through your functions.
```

Comment:
Instead of mentioning `console.log`, I think we'd do a service to learners to just link them to the debugger right away. I'm sure they'll know how to console.log already and it's a huge weakness of learners to not use the debugger.

Let's not make lazy devs.